// just-curves/lib.es2015/internal/cssEasings.js
var c = "cubic-bezier";
var s = "steps";
export var ease = c + "(.25,.1,.25,1)";
export var easeIn = c + "(.42,0,1,1)";
export var easeInBack = c + "(.6,-.28,.735,.045)";
export var easeInCirc = c + "(.6,.04,.98,.335)";
export var easeInCubic = c + "(.55,.055,.675,.19)";
export var easeInExpo = c + "(.95,.05,.795,.035)";
export var easeInOut = c + "(.42,0,.58,1)";
export var easeInOutBack = c + "(.68,-.55,.265,1.55)";
export var easeInOutCirc = c + "(.785,.135,.15,.86)";
export var easeInOutCubic = c + "(.645,.045,.355,1)";
export var easeInOutExpo = c + "(1,0,0,1)";
export var easeInOutQuad = c + "(.455,.03,.515,.955)";
export var easeInOutQuart = c + "(.77,0,.175,1)";
export var easeInOutQuint = c + "(.86,0,.07,1)";
export var easeInOutSine = c + "(.445,.05,.55,.95)";
export var easeInQuad = c + "(.55,.085,.68,.53)";
export var easeInQuart = c + "(.895,.03,.685,.22)";
export var easeInQuint = c + "(.755,.05,.855,.06)";
export var easeInSine = c + "(.47,0,.745,.715)";
export var easeOut = c + "(0,0,.58,1)";
export var easeOutBack = c + "(.175,.885,.32,1.275)";
export var easeOutCirc = c + "(.075,.82,.165,1)";
export var easeOutCubic = c + "(.215,.61,.355,1)";
export var easeOutExpo = c + "(.19,1,.22,1)";
export var easeOutQuad = c + "(.25,.46,.45,.94)";
export var easeOutQuart = c + "(.165,.84,.44,1)";
export var easeOutQuint = c + "(.23,1,.32,1)";
export var easeOutSine = c + "(.39,.575,.565,1)";
export var elegantSlowStartEnd = c + "(.175,.885,.32,1.275)";
export var linear = c + "(0,0,1,1)";
export var stepEnd = s + "(1,0)";
export var stepStart = s + "(1,1)";
