// just-animate/lib/lib/reducers/index.js
Object.defineProperty(exports, "__esModule", { value: true });
var cancel_1 = require("./cancel");
exports.cancel = cancel_1.cancel;
var destroy_1 = require("./destroy");
exports.destroy = destroy_1.destroy;
var finish_1 = require("./finish");
exports.finish = finish_1.finish;
var pause_1 = require("./pause");
exports.pause = pause_1.pause;
var play_1 = require("./play");
exports.play = play_1.play;
var reverse_1 = require("./reverse");
exports.reverse = reverse_1.reverse;
var setup_1 = require("./setup");
exports.setup = setup_1.setup;
var tick_1 = require("./tick");
exports.tick = tick_1.tick;
var update_1 = require("./update");
exports.update = update_1.update;
var update_rate_1 = require("./update-rate");
exports.updateRate = update_rate_1.updateRate;
var update_time_1 = require("./update-time");
exports.updateTime = update_time_1.updateTime;
var append_1 = require("./append");
exports.append = append_1.append;
var insert_1 = require("./insert");
exports.insert = insert_1.insert;
var set_1 = require("./set");
exports.set = set_1.set;
