// just-animate/lib/web/constants.js
Object.defineProperty(exports, "__esModule", { value: true });
var strings_1 = require("../lib/utils/strings");
exports.RUNNING = 'running';
exports.PX = 'px';
exports.DEG = 'deg';
exports.X = 'X';
exports.Y = 'Y';
exports.Z = 'Z';
exports.TRANSLATE = 'translate';
exports.TRANSFORM = 'transform';
exports.transformAngles = strings_1.csvToList('rotateX,rotateY,rotateZ,rotate');
exports.transformScales = strings_1.csvToList('scaleX,scaleY,scaleZ,scale');
exports.transformLengths = strings_1.csvToList('perspective,x,y,z');
exports.transforms = exports.transformAngles.concat(exports.transformScales, exports.transformLengths);
exports.aliases = {
    x: exports.TRANSLATE + exports.X,
    y: exports.TRANSLATE + exports.Y,
    z: exports.TRANSLATE + exports.Z
};
exports.cssLengths = strings_1.csvToList("backgroundSize,border,borderBottom,borderBottomLeftRadius,borderBottomRightRadius,borderBottomWidth,borderLeft,borderLeftWidth,borderRadius,borderRight,borderRightWidth,borderTop,borderTopLeftRadius,borderTopRightRadius,borderTopWidth,borderWidth,bottom,columnGap,columnRuleWidth,columnWidth,columns,flexBasis,font,fontSize,gridColumnGap,gridGap,gridRowGap,height,left,letterSpacing,lineHeight,margin,marginBottom,marginLeft,marginRight,marginTop,maskSize,maxHeight,maxWidth,minHeight,minWidth,outline,outlineOffset,outlineWidth,padding,paddingBottom,paddingLeft,paddingRight,paddingTop,perspective,right,shapeMargin,tabSize,top,width,wordSpacing");
exports.cssProps = [exports.TRANSFORM].concat(exports.transforms, exports.cssLengths);
